File: backend/app/schemas/phase8.py
"""
Pydantic schemas for Phase 8: Personalized Recommendations, Saved Businesses,
Interactions, Natural Language Query, Entrepreneur Profile.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ── Entrepreneur Profile ──────────────────────────────────────────────────────

class EntrepreneurProfileIn(BaseModel):
    detailed_skills:            Optional[str]   = None
    education_level:            Optional[str]   = None
    experience_description:     Optional[str]   = None
    preferred_work_style:       Optional[str]   = None
    daily_available_hours:      Optional[float] = Field(None, ge=0, le=24)
    location_type:              Optional[str]   = None   # rural | semi_urban | urban
    preferred_business_types:   Optional[str]   = None
    family_business_experience: Optional[bool]  = None
    existing_assets:            Optional[str]   = None
    growth_preference:          Optional[str]   = None   # stable | balanced | aggressive
    business_goal:              Optional[str]   = None


class EntrepreneurProfileOut(EntrepreneurProfileIn):
    id:         str
    user_id:    str
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


# ── Semantic match detail ─────────────────────────────────────────────────────

class SemanticMatchDetail(BaseModel):
    semantic_score:    float
    matched_concepts:  List[str]
    explanation:       str
    method:            str


# ── Score breakdown ───────────────────────────────────────────────────────────

class PersonalizedBreakdown(BaseModel):
    semantic_skill:       float
    budget:               float
    market_opportunity:   float
    financial_potential:  float
    experience:           float
    gov_support:          float
    risk:                 float
    interest:             float
    income_goal:          float
    location:             float
    preference_modifier:  float = 0.0


class PersonalizedScoreRaw(BaseModel):
    semantic_skill:       float
    budget:               float
    market_opportunity:   float
    financial_potential:  float
    experience:           float
    gov_support:          float
    risk:                 float
    interest:             float
    income_goal:          float
    location:             float


# ── Recommendation explanation ────────────────────────────────────────────────

class FinancialOutlook(BaseModel):
    min_investment:       float
    max_investment:       float
    monthly_profit_min:   float
    monthly_profit_max:   float
    risk_level:           str


class SemanticMatchSummary(BaseModel):
    score:        float
    concepts:     List[str]
    explanation:  str


class RecommendationExplanation(BaseModel):
    why_recommended:   List[str]
    strengths:         List[str]
    challenges:        List[str]
    next_steps:        List[str]
    financial_outlook: FinancialOutlook
    semantic_match:    SemanticMatchSummary
    disclaimer:        str


# ── Personalized recommendation item ─────────────────────────────────────────

class PersonalizedRecommendationItem(BaseModel):
    rank:           int
    business_id:    str
    business_name:  str
    category:       str
    business_type:  str
    risk_level:     str
    min_investment: float
    max_investment: float
    monthly_profit_min: float
    monthly_profit_max: float
    setup_time_weeks_min: int
    setup_time_weeks_max: int
    suitable_for_rural: bool
    description:    str
    required_skills: str

    final_score:    float
    breakdown:      PersonalizedBreakdown
    raw_scores:     PersonalizedScoreRaw
    semantic_detail: SemanticMatchDetail
    explanation:    RecommendationExplanation

    is_saved:       bool = False
    disclaimer:     str = (
        "⚠️ Estimated figures — actual results depend on execution, local conditions, and market factors."
    )


# ── Personalized recommendation request / response ───────────────────────────

class PersonalizedRecommendationRequest(BaseModel):
    available_capital:    Optional[float] = Field(None, ge=0)
    skills:               Optional[str]   = None
    business_interests:   Optional[str]   = None
    monthly_income_goal:  Optional[float] = Field(None, ge=0)
    preferred_risk:       Optional[str]   = None
    experience_years:     Optional[int]   = Field(None, ge=0)
    location_type:        Optional[str]   = None
    top_n:                int             = Field(8, ge=1, le=15)
    use_preferences:      bool            = True


class PersonalizedRecommendationResponse(BaseModel):
    recommendations:          List[PersonalizedRecommendationItem]
    profile_completeness:     int
    total_businesses_scored:  int
    ai_mode:                  str = "data"   # data | enhanced
    disclaimer:               str = (
        "⚠️ AI-assisted suggestions based on estimated data. "
        "Consult local experts before making business decisions."
    )


# ── Natural language query ────────────────────────────────────────────────────

class NaturalQueryRequest(BaseModel):
    query:          str = Field(..., min_length=3)
    top_n:          int = Field(5, ge=1, le=10)
    use_ai_parsing: bool = True


class ExtractedIntent(BaseModel):
    budget:               Optional[float]
    skills:               Optional[str]
    risk_preference:      Optional[str]
    business_type_hints:  List[str]
    location_type:        Optional[str]
    raw_query:            str


class NaturalQueryResponse(BaseModel):
    recommendations:  List[PersonalizedRecommendationItem]
    extracted_intent: ExtractedIntent
    parse_method:     str
    disclaimer:       str = (
        "⚠️ Natural language parsing is best-effort. "
        "Recommendations are based on estimated data only."
    )


# ── Interaction tracking ──────────────────────────────────────────────────────

class InteractionRequest(BaseModel):
    interaction_type: str = Field(
        ...,
        description="One of: viewed | saved | compared | dismissed | explored"
    )


class InteractionOut(BaseModel):
    id:               str
    business_id:      str
    interaction_type: str
    created_at:       datetime
    model_config = {"from_attributes": True}


class PreferenceSummary(BaseModel):
    preferred_categories: Dict[str, int]
    avoided_categories:   Dict[str, int]
    preferred_risk:       Optional[str]
    total_interactions:   int
    disclaimer:           str = "Preference adjustments are small (+/-10 pts max) and fully transparent."


# ── Saved businesses ──────────────────────────────────────────────────────────

class SavedBusinessIn(BaseModel):
    notes: Optional[str] = Field(None, max_length=500)


class SavedBusinessOut(BaseModel):
    id:          str
    business_id: str
    notes:       Optional[str]
    created_at:  datetime
    model_config = {"from_attributes": True}


class SavedBusinessListOut(BaseModel):
    items: List[SavedBusinessOut]
    total: int

File: frontend/src/types/phase8.ts
/**
 * Phase 8 TypeScript types — Personalized Recommendations, Saved Businesses,
 * Natural Language Query, Interaction Tracking, Entrepreneur Profile.
 */

export interface SemanticMatchDetail {
  semantic_score:    number
  matched_concepts:  string[]
  explanation:       string
  method:            string
}

export interface PersonalizedBreakdown {
  semantic_skill:       number
  budget:               number
  market_opportunity:   number
  financial_potential:  number
  experience:           number
  gov_support:          number
  risk:                 number
  interest:             number
  income_goal:          number
  location:             number
  preference_modifier:  number
}

export interface PersonalizedScoreRaw {
  semantic_skill:       number
  budget:               number
  market_opportunity:   number
  financial_potential:  number
  experience:           number
  gov_support:          number
  risk:                 number
  interest:             number
  income_goal:          number
  location:             number
}

export interface FinancialOutlook {
  min_investment:       number
  max_investment:       number
  monthly_profit_min:   number
  monthly_profit_max:   number
  risk_level:           string
}

export interface SemanticMatchSummary {
  score:        number
  concepts:     string[]
  explanation:  string
}

export interface RecommendationExplanation {
  why_recommended:   string[]
  strengths:         string[]
  challenges:        string[]
  next_steps:        string[]
  financial_outlook: FinancialOutlook
  semantic_match:    SemanticMatchSummary
  disclaimer:        string
}

export interface PersonalizedRecommendationItem {
  rank:               number
  business_id:        string
  business_name:      string
  category:           string
  business_type:      string
  risk_level:         string
  min_investment:     number
  max_investment:     number
  monthly_profit_min: number
  monthly_profit_max: number
  setup_time_weeks_min: number
  setup_time_weeks_max: number
  suitable_for_rural: boolean
  description:        string
  required_skills:    string

  final_score:        number
  breakdown:          PersonalizedBreakdown
  raw_scores:         PersonalizedScoreRaw
  semantic_detail:    SemanticMatchDetail
  explanation:        RecommendationExplanation
  is_saved:           boolean
  disclaimer:         string
}

export interface PersonalizedRecommendationRequest {
  available_capital?:   number
  skills?:              string
  business_interests?:  string
  monthly_income_goal?: number
  preferred_risk?:      string
  experience_years?:    number
  location_type?:       string
  top_n?:               number
  use_preferences?:     boolean
}

export interface PersonalizedRecommendationResponse {
  recommendations:         PersonalizedRecommendationItem[]
  profile_completeness:    number
  total_businesses_scored: number
  ai_mode:                 string
  disclaimer:              string
}

export interface ExtractedIntent {
  budget:               number | null
  skills:               string | null
  risk_preference:      string | null
  business_type_hints:  string[]
  location_type:        string | null
  raw_query:            string
}

export interface NaturalQueryResponse {
  recommendations: PersonalizedRecommendationItem[]
  extracted_intent: ExtractedIntent
  parse_method:    string
  disclaimer:      string
}

export interface PreferenceSummary {
  preferred_categories: Record<string, number>
  avoided_categories:   Record<string, number>
  preferred_risk:       string | null
  total_interactions:   number
  disclaimer:           string
}

export interface SavedBusinessOut {
  id:          string
  business_id: string
  notes:       string | null
  created_at:  string
}

export interface SavedBusinessListOut {
  items: SavedBusinessOut[]
  total: number
}

export interface EntrepreneurProfileIn {
  detailed_skills?:            string
  education_level?:            string
  experience_description?:     string
  preferred_work_style?:       string
  daily_available_hours?:      number
  location_type?:              string
  preferred_business_types?:   string
  family_business_experience?: boolean
  existing_assets?:            string
  growth_preference?:          string
  business_goal?:              string
}

export interface EntrepreneurProfileOut extends EntrepreneurProfileIn {
  id:         string
  user_id:    string
  created_at: string
  updated_at: string
}
